subprojects/core/src/main/java/org/gradle/api/internal/tasks/properties/DefaultTypeMetadataStore.java

```diff
@@ -106,19 +106,29 @@ private static String calculateDisplayName(Iterable<? extends PropertyAnnotation
             Annotation normalizationAnnotation = propertyAnnotations.get(NORMALIZATION);
             Class<? extends Annotation> propertyType = determinePropertyType(typeAnnotation, normalizationAnnotation);
             if (propertyType == null) {
-                validationContext.visitPropertyProblem(WARNING,
-                    propertyAnnotationMetadata.getPropertyName(),
-                    String.format("is not annotated with %s", displayName)
+                validationContext.visitPropertyProblem(problem ->
+                    problem.withId(ValidationProblemId.MISSING_ANNOTATION)
+                        .forProperty(propertyAnnotationMetadata.getPropertyName())
+                        .reportAs(ERROR)
+                        .withDescription(() -> "is missing " + displayName)
+                        .happensBecause("A property without annotation isn't considered during up-to-date checking")
+                        .addPossibleSolution(() -> "Add " + displayName)
+                        .addPossibleSolution("Mark it as @Internal")
+                        .documentedAt("validation_problems", "missing_annotation")
                 );
                 continue;
             }
 
             PropertyAnnotationHandler annotationHandler = propertyAnnotationHandlers.get(propertyType);
             if (annotationHandler == null) {
-                validationContext.visitPropertyProblem(ERROR,
-                    propertyAnnotationMetadata.getPropertyName(),
-                    String.format("is annotated with invalid property type @%s",
-                        propertyType.getSimpleName())
+                validationContext.visitPropertyProblem(problem ->
+                    problem.withId(ValidationProblemId.ANNOTATION_INVALID_IN_CONTEXT)
+                        .forProperty(propertyAnnotationMetadata.getPropertyName())
+                        .reportAs(ERROR)
+                        .withDescription(() -> String.format("is annotated with invalid property type @%s", propertyType.getSimpleName()))
+                        .happensBecause(() -> "The '@" + propertyType.getSimpleName() + "' annotation cannot be used in this context")
```

Comment:
We could list the allowed annotations for the context. Do you think that would be too long?

Reply:
Same problem here: the `validationContext` we get doesn't carry enough information for me to display something useful.

Reply:
`propertyAnnotationHandlers.keySet()` are the allowed annotations, right?

Reply:
Yes you're right! I got confused by the fact I couldn't tell what is the actual context but the allowed annotations are known.